Implement Storage API

thalkz opened the pull request:
## What kind of change does this PR introduce?

This is a first draft for the dart Storage API, with both the Bucket and the Files endpoints.

## Additional context

- I tried to keep the style & comment docs consistent with the the js implementation (https://github.com/supabase/supabase-js)
- I added some basic tests for each endpoint, but they require a working Supabase project (not ideal for tests), accessed using environment variables.
- More advanced testing will be required to make sure everything works fine (other file formats, cacheControl, error handling etc)

This is my first open source contribution, feedback will be much appreciated !

dshukertjr commented on lib/src/storage/storage_file_api.dart:
```diff
@@ -0,0 +1,193 @@
+      if (response.hasError) {
+        return StorageResponse(error: response.error);
+      } else {
+        final signedUrl = '$url${response.data['signedUrl']}';
```
```suggestion
        final signedUrl = '$url${response.data['signedUrl']}';
```

dshukertjr replied:
Hey @thalkz !

Thanks for the amazing work! I am already using this code in one of my app and I love it! I just happened to find a small bug. I believed supabase-js had this bug a while ago as well, but now they seem to use `signedURL`.

thalkz replied:
Hello !
Glad to hear the code works well in your app. I'm updating the PR now

dshukertjr commented on lib/supabase.dart:
```diff
@@ -9,3 +9,4 @@ export 'package:realtime_client/src/realtime_subscription.dart'
 export 'src/auth_user.dart';
```
```suggestion
export 'src/auth_user.dart';
export 'src/storage/types.dart' show FileOptions, SearchOptions, SortBy;
```

dshukertjr replied:
With this users will be able to set `FileOptions` in upload method, `SearchOptions` in `list` method and `SortBy` in declaring `SearchOptions`.

thalkz commented:
I fixed a few things and added the `mocktail` package to mock `fetch` in the tests. All storage tests now work 100% offline. @dshukertjr let me know if you see stuff that should be improved.

thalkz commented:
Not sure if it's an issue or not, but all uploaded files have `content-type: application/octet-stream`. This is the default behavior of `http.MultipartFile`, which is used [here](https://github.com/thalkz/supabase-dart/blob/thalkz/storage/lib/src/storage/fetch.dart#L106). I was thinking about using the [mime package](https://pub.dev/packages/mime) to lookup the mime type of files, using their filename.

dshukertjr commented:
It looks amazing! @phamhieu what do you think?

phamhieu commented:
Wow, this' amazing @thalkz. Thank you.

thalkz commented:
Thanks ! Happy to contribute
